### PR TITLE
Added tcp fragmentation for freedom outbound

### PR DESCRIPTION
**Pull request by sambali9**
This is inspired by https://github.com/GFW-knocker/gfw_resist_tls_proxy
Usage:
```
{
    "protocol": "freedom",
    "settings": {
        "fragment": {
            "delay": 100, // Delay after each fragment
            "divider": 10, // Number of packets to send, if the original packet size is not divisible by the divider the last packet will be fragment + remainder
            "packetNumber": 1 // nth packet to fragment, if 0 all outbound packets will be fragmented
        }
    }
}
```

### CODE REVIEW

**Comment by yuhan6665** [comment]
@RPRX

**Comment by RPRX** [comment]
感谢 PR，建议如下：

1. 需设置 TCP_NODELAY
2. delay 改名为 interval，可以设置一个范围，每次随机，比如 `"interval": "100-300"`（与 REALITY 的 spiderX 类似）
3. 或许将策略改为指定每次发送的长度更好？比如 `"50-100"`，每次发送 50-100 bytes

**Comment by sambali9** [comment]
Usage:
```
{
    "protocol": "freedom",
    "settings": {
        "fragment": {
           "length": "10-20", // length of each fragment to send it can be single value or range it has to be a string. If it is a range the values are randomly chosen.
            "interval": "100-500", // the time to wait for each packet can be single value or range but it has to be a string. If it is a range the values are randomly chosen.
            "packetNumber": 1 // nth packet to fragment, if 0 all outbound packets will be fragmented
        }
    },
    "streamSettings": {
        "sockopt": {
            "tcpNoDelay": true // This needs to be enabled so the kernel doesn't cache the fragments
        }
    }
}
```

**Comment by RPRX** [comment]
感谢，既然另外两个参数都是 `"x-y"` 的形式了，我们也可以设置 `"packets": "1-3"`，它的含义为前三个包，这样更灵活一些

此项为 `""` 时，all outbound packets will be fragmented

此外我想知道，目前这个 PR 的代码在伊朗是否有效，比如可以直接上 youtube？

**Comment by RPRX** [comment]
`"packets": "0-3"` 这种直接报错

**Comment by sambali9** [comment]
In Iran the censorship system is not centralised and results can vary from different ISPs. For me I couldn't get `www.youtube.com` to work but it worked with sites behind Cloudflare like `www.clubhouseapi.com` which was blocked without fragmentation. The setting that worked:
```
{
            "protocol": "freedom",
            "settings": {
                "fragment": {
                    "length": "1-10",
                    "interval": "10-100",
                    "packetNumber": 1
                }
            },
            "streamSettings": {
                "sockopt": {
                    "tcpNoDelay": true
                }
            }
}
```

I didn't understand what you meant with `"packets": "1-3"`, should this match the first, second and the third packet or is it gonna be like go slices where this matches the second and the third packets?

**Comment by RPRX** [comment]
> I didn't understand what you meant with `"packets": "1-3"`, should this match the first, second and the third packet or is it gonna be like go slices where this matches the second and the third packets?

The first, second and the third packet.

**Comment by RPRX** [comment]
The order should be like:

```go
type Fragment struct {
	Packets  string `json:"packets"`
	Length   string `json:"length"`
	Interval string `json:"interval"`
}
```

**Comment by mortada-jafar** [comment]
> `"packets": "0-3"`This kind of direct error

I tested it not work!

Proider : MCI

**Comment by sambali9** [comment]
>dirty IP: connection is established but the upload speed has been throttled.
 
Do you get good result with python scripts at [gfw_resist_tls_proxy](https://github.com/GFW-knocker/gfw_resist_tls_proxy)?
If you get good result please provide your configurations (packet length and delay time).


آپلودتون با اسکریپت های [gfw_resist_tls_proxy](https://github.com/GFW-knocker/gfw_resist_tls_proxy) خوبه؟
اگر آپلودتون با نسخه پایتون خوب هست لطفا مقادیر مثل اندازه پکت ها و زمان بینشون رو بنویسید.

**Comment by mortada-jafar** [comment]
> 
same as [gfw_resist_tls_proxy](https://github.com/GFW-knocker/gfw_resist_tls_proxy)
خیر الان تست کردم فرقی ندارد

**Comment by hiddify1** [comment]
It is great. it would be greater if you add it for all outbounds not only the freedom outbound

**Comment by sambali9** [comment]
> It is great. it would be greater if you add it for all outbounds not only the freedom outbound

There is actually no need to do that. You can simply use the `dialerProxy` to do with other outbounds:
```
"outbounds": [
        {
            "protocol": "vmess",
            "settings": {
                "vnext": [
                    {
                        "address": "1.2.3.4",
                        "port": 1234,
                        "users": [
                            {
                                "id": "5e0fa60c-0074-4ca7-87c2-3a1af806e89e",
                                "level": 0
                            }
                        ]
                    }
                ]
            },
            "streamSettings": {
                "network": "tcp",
                "sockopt": {
                    "dialerProxy": "fragment"
                }
            },
            "tag": "proxy"
        },
        {
            "protocol": "freedom",
            "settings": {
                "fragment": {
                    "length": "1",
                    "interval": "100-150",
                    "packets": "1-3"
                }
            },
            "streamSettings": {
                "sockopt": {
                    "tcpNoDelay": true
                }
            },
            "tag": "fragment"
        }
    ]
```

**Comment by sambali9** [comment]
@RPRX BTW in case you didn't notice I have changed the packetNumber to packets as requested

**Comment by sambali9** [comment]
Usage:
```
{
    "protocol": "freedom",
    "settings": {
        "fragment": {
           "length": "10-20", // length of each fragment to send in bytes. It can be single value or range it has to be a string. If it is a range the values are randomly chosen.
            "interval": "100-500", // the time to wait for each packet in milliseconds. Can be single value or range but it has to be a string. If it is a range the values are randomly chosen.
            "packets": "1-3" // nth packet(s) to fragment, if empty all outbound packets will be fragmented, can be a single number or a range. The start packet cannot be 0 so 0-3 is not acceptable. The range numbers are inclusive.
        }
    },
    "streamSettings": {
        "sockopt": {
            "tcpNoDelay": true // This needs to be enabled so the kernel doesn't cache the fragments
        }
    }
}
```

File: proxy/freedom/freedom.go [hunk 1]
```diff
@@ -324,3 +341,61 @@ func (w *PacketWriter) WriteMultiBuffer(mb buf.MultiBuffer) error {
...
+	}
+
+	if len(buf) <= w.minLength {
+		return w.Writer.Write(buf)
+	}
+	nTotal := 0
+	for {
+		if nTotal >= len(buf) {
+			return nTotal, nil
+		}
```
Comment (RPRX) [inline proxy/freedom/freedom.go:369]: 移至 randomInterval 前

File: proxy/freedom/freedom.go [hunk 2]
```diff
@@ -324,3 +341,61 @@ func (w *PacketWriter) WriteMultiBuffer(mb buf.MultiBuffer) error {
...
+		randomByteSize := int(randBetween(int64(w.minLength), int64(w.maxLength)))
+		if nTotal+randomByteSize > len(buf) {
+			n, err := w.Writer.Write(buf[nTotal:])
+			if err != nil {
```
Comment (RPRX) [inline proxy/freedom/freedom.go:373]: 直接修改 randomByteSize，和下面的代码合并

File: proxy/freedom/freedom.go [hunk 3]
```diff
@@ -324,3 +341,61 @@ func (w *PacketWriter) WriteMultiBuffer(mb buf.MultiBuffer) error {
...
+	startPacket int
+	endPacket   int
+	PacketCount int
+}
+
+func (w *FragmentWriter) Write(buf []byte) (int, error) {
+	w.PacketCount += 1
+	if !((w.PacketCount >= w.startPacket && w.PacketCount <= w.endPacket) || w.startPacket == 0) {
+		return w.Writer.Write(buf)
+	}
```
Comment (RPRX) [inline proxy/freedom/freedom.go:360]: 不需要判断 w.startPacket

File: proxy/freedom/freedom.go [hunk 4]
```diff
@@ -324,3 +341,61 @@ func (w *PacketWriter) WriteMultiBuffer(mb buf.MultiBuffer) error {
...
+
+	if len(buf) <= w.minLength {
+		return w.Writer.Write(buf)
+	}
```
Comment (RPRX) [inline proxy/freedom/freedom.go:364]: 与上面的判断合并

File: proxy/freedom/freedom.go [hunk 5]
```diff
@@ -324,3 +341,61 @@ func (w *PacketWriter) WriteMultiBuffer(mb buf.MultiBuffer) error {
...
+
+// stolen from github.com/xtls/xray-core/transport/internet/reality
+func randBetween(left int64, right int64) int64 {
+	if left == right {
+		return left
+	}
+	// swap left and right if left > right so we always get a positive value
+	if left > right {
+		left, right = right, left
+	}
```
Comment (RPRX) [inline proxy/freedom/freedom.go:398]: 移至配置解析处

**Comment by RPRX** [comment]
感谢！以上是一些使代码逻辑更清晰的建议

**Comment by RPRX** [comment]
合并这个 PR 后，计划给 WSS 的 path 添加一些参数以支持快速配置与分享 https://github.com/XTLS/Xray-core/issues/2000#issuecomment-1528913218

**Comment by A5DkjGQUZx** [comment]
Great work... KUDOS TO ALL

**Comment by weareomid** [comment]
Why is it not merged yet?

**Comment by A5DkjGQUZx** [comment]
> Why is it not merged yet?

You can try binaries with the new features here: https://github.com/XTLS/Xray-core/actions/runs/4901135431

How to config?! https://github.com/XTLS/Xray-core/pull/2021#issuecomment-1536006968

**Comment by modaresimr** [comment]
Hi
I think there is a problem in the implementation

I could not bypass blocking using it while it works perfectly with the alternative approaches.

Since your work is really great it would be great if you can fix the issue or would you please send me a config that works perfectly in Iran?

**Comment by RPRX** [comment]
感谢 PR，考虑到伊朗的情况，v1.8.2 将包含这个 PR 的内容

**Comment by WordsWorthLess** [comment]
> > Why is it not merged yet?
> 
> You can try binaries with the new features here: https://github.com/XTLS/Xray-core/actions/runs/4901135431
> 
> How to config?! [#2021 (comment)](https://github.com/XTLS/Xray-core/pull/2021#issuecomment-1536006968)

Tried it with websocket/gRPC with CloudFlare CDN
config:
```
		{
			"tag": "ws_fragmentation",
			"sendThrough": "::",
			"protocol": "vless",
			"settings": {
				"vnext": [
					{
						"address": "cloudflare_ip",
						"port": 443,
						"users": [
							{
								"id": "uuid",
								"encryption": "none"
							}
						]
					}
				]
			},
			"streamSettings": {
				"network": "ws",
				"security": "tls",
				"tlsSettings": {
					"serverName": "server.name",
					"alpn": [
						"http/1.1"
					],
					"allowInsecure": false
				},
				"wsSettings": {
					"path": "/wspath"
				},
				"sockopt": {
					"domainStrategy": "AsIs",
					"dialerProxy": "fragment",
					"mark": 255
				}
			}
		},
		{
			"tag": "fragment",
			"protocol": "freedom",
			"settings": {
				"fragment": {
					"length": "8-15",
					"interval": "50-250",
					"packets": "1-3"
				}
			},
			"streamSettings": {
				"sockopt": {
					"tcpNoDelay": true,
					"mark": 255
				}
			}
```
And  I got TCP RST
fragment1
fragment2

**Comment by WordsWorthLess** [comment]
> > Why is it not merged yet?
> 
> You can try binaries with the new features here: https://github.com/XTLS/Xray-core/actions/runs/4901135431
> 
> How to config?! [#2021 (comment)](https://github.com/XTLS/Xray-core/pull/2021#issuecomment-1536006968)
> > Why is it not merged yet?
> 
> You can try binaries with the new features here: https://github.com/XTLS/Xray-core/actions/runs/4901135431
> 
> How to config?! [#2021 (comment)](https://github.com/XTLS/Xray-core/pull/2021#issuecomment-1536006968)

Tried it with websocket/gRPC with CloudFlare CDN
config:
```
		{
			"tag": "ws_fragmentation",
			"sendThrough": "::",
			"protocol": "vless",
			"settings": {
				"vnext": [
					{
						"address": "cloudflare_ip",
						"port": 443,
						"users": [
							{
								"id": "uuid",
								"encryption": "none"
							}
						]
					}
				]
			},
			"streamSettings": {
				"network": "ws",
				"security": "tls",
				"tlsSettings": {
					"serverName": "server.name",
					"alpn": [
						"http/1.1"
					],
					"allowInsecure": false
				},
				"wsSettings": {
					"path": "/wspath"
				},
				"sockopt": {
					"domainStrategy": "AsIs",
					"dialerProxy": "fragment",
					"mark": 255
				}
			}
		},
		{
			"tag": "fragment",
			"protocol": "freedom",
			"settings": {
				"fragment": {
					"length": "8-15",
					"interval": "50-250",
					"packets": "1-3"
				}
			},
			"streamSettings": {
				"sockopt": {
					"tcpNoDelay": true,
					"mark": 255
				}
			}
```
And  I got TCP RST

**Comment by A5DkjGQUZx** [comment]
> You can try binaries with the new features here: https://github.com/XTLS/Xray-core/actions/runs/4901135431
> 
> How to config?! [#2021 (comment)](https://github.com/XTLS/Xray-core/pull/2021#issuecomment-1536006968)

You can try binaries with the new features here: https://github.com/XTLS/Xray-core/actions/runs/4901135431
How to config?! https://github.com/XTLS/Xray-core/pull/2021#issuecomment-1536006968

**Comment by ahmadrezaei** [comment]
How we can use it in Dokodemo-Door ?

**Pull request merged**